Added calls to API to store/get offsets

BScong opened the pull request:
Fixes #56 

As the backend only works for one type (only one storage per item_id), I only store offsets for full articles (not summaries).

I store the offset when the user pauses the content. Te next time the user plays the content, it will begin from this offset.
At the end of the article, I reset the offset to 0.

dylano commented:
Noticed in testing: the offset does not get updated if I am listening to one article and then directly jump into another one without an explicit 'stop' or 'pause' command:
```
me: alexa, ask scout to play the article about football.
alexa: summary or full?
me: full article
alexa: <starts playing football>
me: alexa, ask scout to play the article about poodles
alexa: summary or full?
me: full article
alexa: <starts playing poodles>
```

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -74,10 +74,21 @@ var state_handlers = {
+              0
+            )
+            .catch(function(err) {
+              console.log('Error during offset update.');
```
In catch blocks like this, dump the contents of `err` too so we can get some idea of what went wrong.

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -377,6 +399,74 @@ var scout_agent = (function() {
+    },
+    getArticleStatus: function(userId, articleId) {
+      return new Promise((resolve, reject) => {
+        console.log('getArticleStatus: ' + articleId + ' for ' + userId);
```
Consider template literals for all of these string constructions -- they can be easier to read (and easier to type!):
```
console.log(`getArticleStatus: ${articleId} for ${userId}`);
```

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -377,6 +399,74 @@ var scout_agent = (function() {
+            userId +
+            '/' +
+            articleId,
+          method: 'GET',
```
```
`http://${process.env.SCOUT_ADDR}/article-status/${userId}/...`
```

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -521,21 +612,36 @@ function synthesisHelperUrl(stateObj) {
       );
+    }
+
+    Promise.all([getArticle, getOffset])
```
These variable names are a bit confusing -- especially in JS where functions can get passed around in parameters like this, I'd stay away from including verbs in your object names. Just going with `article` and `offset` would read better.

BScong commented:
I added the support for the case when we ask for another article while listening.

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -115,6 +126,22 @@ var state_handlers = {
+            this.attributes['offsetInMilliseconds']
+          )
+          .catch(function(err) {
+            console.log('Error during offset update.');
```
Another catch which is hiding error details -- take a look at this and the other ones in the file. If this is an unexpected error, try to provide us some more details when we are troubleshooting this via log files somewhere down the line.

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -377,6 +404,70 @@ var scout_agent = (function() {
+            offset
+        );
+        let scoutOptions = {
+          uri: 'http://' + process.env.SCOUT_ADDR + '/article-status/',
```
I didn't flag all of the url/string construction places in the previous commit, I was just calling out a couple of examples but now this is inconsistent with the template literals you've switched to above. You can do it either way; I find the templates more readable but it's a personal preference thing. In any case, let's not mix both methods in the same PR. Sorry for the detour.

BScong replied:
Yeah, I forgot that I had the same syntax elsewhere, thanks for noticing!

dylano commented on lambda/custom/state_handlers.js:
```diff
@@ -115,6 +126,22 @@ var state_handlers = {
       audio_controller.stop.call(this);
+      this.emit('StoppedArticle');
+    },
+    StoppedArticle: function() {
```
👍 This is working great now.

BScong commented:
Should be fine now. Thanks.

As we just landed the PR on scout-ua (https://github.com/MozScout/scout-ua/pull/102) to add the offset field to the other endpoints (/article and /intent:SearchAndPlay). I think it's better to fetch the offset from /article than do a request to /article-status. I'm going to implement this so I moving this back to 'In Progress'

dylano commented:
There shouldn't be much of a difference...? In either case I think you do want to check the current value every time just before you start playing (in case the user is going back & forth from Alexa to mobile) so I'm not seeing the benefit in moving away from article-status.

BScong commented:
Alexa is doing a call to /article to fetch the article audio file url, and is doing at the same time a call to /article-status to fetch the offset. But since /article also returns the offset (added with the scout-ua PR), the call to /article-status is useless ?

dylano commented:
Yeah, in that case it's not useful to make the article-status call. I was kind of thinking of a pause/resume situation where they might have listened to the first part on alexa, then switched to the app for a while, then came back to alexa to finish. I'm just wanting to make sure that alexa is always picking up at the right offset.

BScong commented:
Updated. Removed the getArticleStatus method as we are now fetching the offset from /article or /summary. We only make the article-status call to update the offset now.
